pkg/daemon/update.go

```diff
@@ -1503,22 +1503,27 @@ func (dn *Daemon) writeUnits(units []ign3types.Unit) error {
 			glog.V(2).Infof("Created symlink unit %q to %s", u.Name, pathDevNull)
 
 			continue
-		}
-
-		if u.Contents != nil && *u.Contents != "" {
+		} else if u.Contents != nil && *u.Contents != "" {
 			glog.Infof("Writing systemd unit %q", u.Name)
 			if _, err := os.Stat("/usr" + fpath); err == nil &&
 				dn.os.IsCoreOSVariant() {
 				if err := createOrigFile("/usr"+fpath, fpath); err != nil {
 					return err
 				}
 			}
-			// write the unit to disk
+			// write the unit to fpath
 			if err := writeFileAtomicallyWithDefaults(fpath, []byte(*u.Contents)); err != nil {
 				return fmt.Errorf("failed to write systemd unit %q: %v", u.Name, err)
 			}
 
 			glog.V(2).Infof("Successfully wrote systemd unit %q: ", u.Name)
+		} else {
```

Comment:
Are there cases we need to leave something there? Or create an orig file or something?

Reply:
So, let's say we got here, we just know that mask is nil and contents is nil, if I'm reading this correctly.

A user could have created dropins for an existing service, for example. It doesn't mean we should clean up the unit. If anything I don't think this should be gated on the else

Reply:
I should probably reword the comment, but as far as I can tell dropins go into `u.Name+".d"` rather than `u.Name`, which is what I'm removing here. Is there a case where mask is nil and contents is nil, but there should still be something in the file `u.Name`?

Reply:
An example: enabling or disabling a service that already exists on the system. To do so you can simply pass a machineconfig that contains that service, with no mask and no content, and it should enable. Example spec:

```
spec:
  config:
    ignition:
      version: 3.1.0
    systemd:
      units:
        - name: test.service
          enabled: true
```

The main issue with our assumptions here is that all the systemd services are fully defined via machineconfigs, which is not the case. There are many that are shipped by default on an RHCOS system, for example

Reply:
Ok I was assuming those were presets, but I guess that's not true

Reply:
How can we unmask something then? Masking overwrites everything at `fpath`, and I don't see a backup. If it's not a preset, how would we restore `fpath` when unmasking?

Reply:
After discussing with Matthew, we think this actually should work. Although the current logic where if we have a unit with no mask and no content (e.g. enable/dropin) still feels minorly... incorrect and inefficient, although I can't think of a scenario that would cause issues.

One option if we want to make absolutely sure is instead of this as an else condition, just have it be an if condition that handles only mask=false, so we aren't "removing" something that isn't there on e.g. an enable operation, saving that extra RemoveAll call when the user did not define mask=false. For masks that are deleted straight up the deleteStaleData will handle it as usual. WDYT @mkenigs

Reply:
A case that misses is `mask: true` -> `enabled: true`. Since both mask and content are nil, `fpath` won't get touched, and the unit would still be masked

Reply:
Another bug (I think):
`content: someContent` -> `enabled: true`. Since content is nil for the second config, `fpath` doesn't get removed, and `someContent` still applies rather than whatever ships with RHCOS

Reply:
I think we either need to accept buggy/weird behavior or use both the old units and the new units to decide what action to take